From: Joel Selvaraj <foss@joelselvaraj.com>
Date: Thu, 2 Oct 2025 22:39:12 -0500
Subject: Revert "pmdomain: core: Leave powered-on genpds on until
 late_initcall_sync"

This reverts commit 13a4b7fb62600e1c0738fdb0b7176555ff05aadf.
---
 drivers/pmdomain/core.c   | 10 ++--------
 include/linux/pm_domain.h |  1 -
 2 files changed, 2 insertions(+), 9 deletions(-)

diff --git a/drivers/pmdomain/core.c b/drivers/pmdomain/core.c
index 18951ed..5cef6de 100644
--- a/drivers/pmdomain/core.c
+++ b/drivers/pmdomain/core.c
@@ -931,12 +931,11 @@ static void genpd_power_off(struct generic_pm_domain *genpd, bool one_dev_on,
 	 * The domain is already in the "power off" state.
 	 * System suspend is in progress.
 	 * The domain is configured as always on.
-	 * The domain was on at boot and still need to stay on.
 	 * The domain has a subdomain being powered on.
 	 */
 	if (!genpd_status_on(genpd) || genpd->prepared_count > 0 ||
 	    genpd_is_always_on(genpd) || genpd_is_rpm_always_on(genpd) ||
-	    genpd->stay_on || atomic_read(&genpd->sd_count) > 0)
+	    atomic_read(&genpd->sd_count) > 0)
 		return;
 
 	/*
@@ -1347,12 +1346,8 @@ static int __init genpd_power_off_unused(void)
 	pr_info("genpd: Disabling unused power domains\n");
 	mutex_lock(&gpd_list_lock);
 
-	list_for_each_entry(genpd, &gpd_list, gpd_list_node) {
-		genpd_lock(genpd);
-		genpd->stay_on = false;
-		genpd_unlock(genpd);
+	list_for_each_entry(genpd, &gpd_list, gpd_list_node)
 		genpd_queue_power_off_work(genpd);
-	}
 
 	mutex_unlock(&gpd_list_lock);
 
@@ -2357,7 +2352,6 @@ int pm_genpd_init(struct generic_pm_domain *genpd,
 	INIT_WORK(&genpd->power_off_work, genpd_power_off_work_fn);
 	atomic_set(&genpd->sd_count, 0);
 	genpd->status = is_off ? GENPD_STATE_OFF : GENPD_STATE_ON;
-	genpd->stay_on = !is_off;
 	genpd->sync_state = GENPD_SYNC_STATE_OFF;
 	genpd->device_count = 0;
 	genpd->provider = NULL;
diff --git a/include/linux/pm_domain.h b/include/linux/pm_domain.h
index 78ebca5..1590c99 100644
--- a/include/linux/pm_domain.h
+++ b/include/linux/pm_domain.h
@@ -205,7 +205,6 @@ struct generic_pm_domain {
 	unsigned int performance_state;	/* Aggregated max performance state */
 	cpumask_var_t cpus;		/* A cpumask of the attached CPUs */
 	bool synced_poweroff;		/* A consumer needs a synced poweroff */
-	bool stay_on;			/* Stay powered-on during boot. */
 	enum genpd_sync_state sync_state; /* How sync_state is managed. */
 	int (*power_off)(struct generic_pm_domain *domain);
 	int (*power_on)(struct generic_pm_domain *domain);
